UFDR Error when was save with protection

I received an .ufdr file and when processing it with IPED I got the error below.

iped --portable -d "k:\xxx.ufdr" -o "i:\xxx"
2026-02-20 16:15:05     [ERROR] [app.processing.Main]                   Processing Error:
java.lang.Exception: Error decoding datasource k:\xxx.ufdr
        at iped.engine.datasource.ItemProducer.run(ItemProducer.java:153) ~[iped-engine-4.3.0.jar:?]
Caused by: org.xml.sax.SAXParseException: Content is not allowed in prolog.
        at org.apache.xerces.parsers.AbstractSAXParser.parse(Unknown Source) ~[xercesImpl-2.12.2.jar:?]
        at org.apache.xerces.jaxp.SAXParserImpl$JAXPSAXParser.parse(Unknown Source) ~[xercesImpl-2.12.2.jar:?]
        at iped.engine.datasource.UfedXmlReader.read(UfedXmlReader.java:272) ~[iped-engine-4.3.0.jar:?]
        at iped.engine.datasource.UfedXmlReader.read(UfedXmlReader.java:251) ~[iped-engine-4.3.0.jar:?]
        at iped.engine.datasource.ItemProducer.run(ItemProducer.java:123) ~[iped-engine-4.3.0.jar:?]

Then I opened the file with 7z (since ufdr = zip) and started looking at the folder structure, as shown in the images below.

<img width="1459" height="71" alt="image" src="https://github.com/user-attachments/assets/988310bb-5eab-42f1-afe3-c630cd8f7fbc" />

Inside the folders we have:
<img width="683" height="220" alt="image" src="https://github.com/user-attachments/assets/738ae1dc-f817-4fb5-8b25-41078dbaf517" />

Inside the folders, all the files have normal names, but when you try to open them, they are all corrupted.

Then I opened the file with celebrite, which showed that it was a Google file, apparently sent by Google itself, and that someone had converted it to ufder using celebrite.
<img width="540" height="312" alt="image" src="https://github.com/user-attachments/assets/f1a0e46d-877e-4e61-8d08-79b85fb2a154" />

While browsing through it, I can see all the files.

Apparently they are encrypted and only Cellebrite knows how to read them, which explains the corrupted XML IPED error.

After doing some research, I discovered the problem with IPED not being able to index this UFDR; I tested it with several other files in the same situation.
The problem occurs because whoever generated the .ufdr file, which is generated on the Celebritte screen, in the "report" option, sub-option generate report, clicked on the option shown below on the screen.

<img width="681" height="352" alt="image" src="https://github.com/user-attachments/assets/171ded62-07bc-4d0e-9d10-46c5bb994458" />

This generated a protected ufdr file, but in practice it's useless, since you can simply open the file with Cellebrite and everything will be there.

To resolve the issue, after reopening the file with Cellebrite, simply go to the same report generation option and leave the option blank. This will create a new, unprotected .ufdr file, and IPED will then function normally.

Additionally, I observed that all ufdr files contain the Report.xml file, and comparing the first 16 bytes, they all have the same sequence below.
<img width="238" height="33" alt="image" src="https://github.com/user-attachments/assets/df488be4-2d35-4725-a97a-19ffa4748875" />
Would it be worthwhile for IPED to verify this and, in this situation, inform the user on how to resolve it?

@aberenguel analyzed with these results:
https://github.com/sepinf-inc/IPED/discussions/2801#discussioncomment-15891666
https://github.com/sepinf-inc/IPED/discussions/2801#discussioncomment-15891701


paulobreim

--- a/iped-engine/src/main/java/iped/engine/datasource/UfedXmlReader.java
+++ b/iped-engine/src/main/java/iped/engine/datasource/UfedXmlReader.java
@@ -6,7 +6,6 @@ import java.io.File;
 import java.io.FileInputStream;
 import java.io.IOException;
 import java.io.InputStream;
-import java.io.InputStreamReader;
 import java.io.OutputStreamWriter;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
@@ -39,6 +38,8 @@ import java.util.stream.Collectors;
 import javax.xml.parsers.SAXParser;
 import javax.xml.parsers.SAXParserFactory;
 
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.FilenameUtils;
 import org.apache.commons.lang3.StringUtils;
 import org.apache.logging.log4j.Level;
 import org.apache.logging.log4j.LogManager;
@@ -99,6 +100,12 @@ public class UfedXmlReader extends DataSourceReader {
     private final Level CONSOLE = Level.getLevel("MSG"); //$NON-NLS-1$
 
     private static final String[] HEADER_STRINGS = { "project id", "extractionType", "sourceExtractions" }; //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$
+    private static final byte[] UFDR_PROTECTION_NO_PASSWORD_REPORT_XML_INITIAL_BYTES = new byte[] { 0x2D, 0x06, 0x52, 0x6D };
+    private static final int UFDR_REPORT_XML_INITIAL_BYTES_TO_READ = 1024;
+
+    private static final String UFDR_EXTENSION = "ufdr";
+    private static final String XML_REPORT_EXTENSION = "xml";
+
 
     private static final String AVATAR_PATH_META = ExtraProperties.UFED_META_PREFIX + "contactphoto_extracted_path"; //$NON-NLS-1$
     private static final String ATTACH_PATH_META = ExtraProperties.UFED_META_PREFIX + "attachment_extracted_path"; //$NON-NLS-1$
@@ -152,41 +159,40 @@ public class UfedXmlReader extends DataSourceReader {
     @Override
     public boolean isSupported(File datasource) {
 
-        if (datasource.getName().toLowerCase().endsWith(".ufdr")) {
+        if (FilenameUtils.isExtension(datasource.getName(), UFDR_EXTENSION)) {
             return true;
         }
 
-        InputStream xmlReport = lookUpXmlReportInputStream(datasource);
-        IOUtil.closeQuietly(xmlReport);
-
-        if (xmlReport != null)
-            return true;
-
-        return false;
+        // supports any folder with valid XML report inside
+        try (InputStream xmlReport = lookUpXmlReportInputStream(datasource)) {
+            return xmlReport != null;
+        } catch (Exception e) {
+            return false;
+        }
     }
 
     private InputStream getXmlInputStream(File file) {
-        if (file.getName().toLowerCase().endsWith(".xml")) { //$NON-NLS-1$
-            try (InputStreamReader reader = new InputStreamReader(new FileInputStream(file), "UTF-8")) { //$NON-NLS-1$
-                char[] cbuf = new char[1024];
-                int off = 0, i = 0;
-                while (off < cbuf.length && (i = reader.read(cbuf, off, cbuf.length - off)) != -1)
-                    off += i;
-                String header = new String(cbuf, 0, off);
-                for (String str : HEADER_STRINGS)
-                    if (!header.contains(str))
-                        return null;
+        if (FilenameUtils.isExtension(file.getName(), XML_REPORT_EXTENSION)) {
+            try (InputStream fis = new FileInputStream(file)) {
+                byte[] initialBytes = fis.readNBytes(UFDR_REPORT_XML_INITIAL_BYTES_TO_READ);
+                if (!containsHeaderStrings(initialBytes)) {
+                    return null;
+                }
+
                 return new FileInputStream(file);
 
             } catch (IOException e) {
                 throw new RuntimeException(e);
             }
-        } else if (file.getName().toLowerCase().endsWith(".ufdr")) {
+        } else if (FilenameUtils.isExtension(file.getName(), UFDR_EXTENSION)) {
             try {
                 ufdrFile = file;
                 String xml = "report.xml";
                 if (!getUISF().entryExists(xml)) {
                     xml = "Report.xml";
+                    if (!getUISF().entryExists(xml)) {
+                        return null;
+                    }
                 }
                 return getUISF().getSeekableInputStream(xml);
 
@@ -195,6 +201,21 @@ public class UfedXmlReader extends DataSourceReader {
             }
         }
         return null;
+    }
+
+    private boolean isReportXmlProtected(byte[] initialBytes) {
+        int len = UFDR_PROTECTION_NO_PASSWORD_REPORT_XML_INITIAL_BYTES.length;
+        return Arrays.equals(initialBytes, 0, len, UFDR_PROTECTION_NO_PASSWORD_REPORT_XML_INITIAL_BYTES, 0, len);
+    }
+
+    private boolean containsHeaderStrings(byte[] initialBytes) {
+        String header = new String(initialBytes, StandardCharsets.UTF_8);
+        for (String str : HEADER_STRINGS)
+            if (header.contains(str))
+                return true;
+
+        return false;
+
     }
 
     private boolean entryExists(String entryPath) {
@@ -233,16 +254,13 @@ public class UfedXmlReader extends DataSourceReader {
     private InputStream lookUpXmlReportInputStream(File root) {
         if (root.isFile())
             return getXmlInputStream(root);
-        File[] files = root.listFiles();
-        if (files != null) {
-            for (File file : files) {
-                if (file.getName().toLowerCase().endsWith(".xml")) {
-                    InputStream is = getXmlInputStream(file);
-                    if (is != null)
-                        return is;
-                }
-            }
+
+        for (File file : FileUtils.listFiles(root, new String[]{XML_REPORT_EXTENSION}, false)) {
+            InputStream is = getXmlInputStream(file);
+            if (is != null)
+                return is;
         }
+
         return null;
     }
 
@@ -261,6 +279,8 @@ public class UfedXmlReader extends DataSourceReader {
         try {
             xmlStream = lookUpXmlReportInputStream(root);
 
+            validateXmlStream(xmlStream);
+
             configureParsers();
 
             SAXParserFactory spf = SAXParserFactory.newInstance();
@@ -272,6 +292,33 @@ public class UfedXmlReader extends DataSourceReader {
             xmlReader.parse(new InputSource(new UFEDXMLWrapper(xmlStream)));
         } finally {
             IOUtil.closeQuietly(xmlStream);
+        }
+    }
+
+    private void validateXmlStream(InputStream xmlStream) throws IOException {
+        if (xmlStream == null) {
+            if (root.isFile()) {
+                throw new RuntimeException("Invalid UFDR file: XML report not found. File: " + root.getAbsolutePath());
+            } else {
+                throw new RuntimeException("Invalid UFDR folder: No XML report has been found. Folder: " + root.getAbsolutePath());
+            }
+        }
+
+        if (root.isFile()) {
+            xmlStream.mark(0);
+            byte[] initialBytes = xmlStream.readNBytes(UFDR_REPORT_XML_INITIAL_BYTES_TO_READ);
+
+            if (isReportXmlProtected(initialBytes)) {
+                throw new RuntimeException(
+                        "Unsupported UFDR file: protection is enabled. Generate the UFDR file again with protection disabled. File: "
+                                + root.getAbsolutePath());
+            }
+
+            if (!containsHeaderStrings(initialBytes)) {
+                throw new RuntimeException("Invalid UFDR file: XML report is not valid. "
+                        + "Protection may be enabled. Generate the UFDR file again with protection disabled. File: " + root.getAbsolutePath());
+            }
+            xmlStream.reset();
         }
     }
 
@@ -317,7 +364,7 @@ public class UfedXmlReader extends DataSourceReader {
         rootItem.setDataSource(evidenceSource);
         rootItem.setIdInDataSource("");
         rootItem.setHasChildren(true);
-        if (root.getName().endsWith(".ufdr")) {
+        if (FilenameUtils.isExtension(root.getName(), UFDR_EXTENSION)) {
             rootItem.setLength(root.length());
             rootItem.setSumVolume(false);
         }

--- a/iped-engine/src/main/java/iped/engine/datasource/UfedXmlReader.java
+++ b/iped-engine/src/main/java/iped/engine/datasource/UfedXmlReader.java
@@ -164,11 +164,9 @@ public class UfedXmlReader extends DataSourceReader {
         }
 
         // supports any folder with valid XML report inside
-        try (InputStream xmlReport = lookUpXmlReportInputStream(datasource)) {
-            return xmlReport != null;
-        } catch (Exception e) {
-            return false;
-        }
+        InputStream xmlReport = lookUpXmlReportInputStream(datasource);
+        IOUtil.closeQuietly(xmlReport);
+        return xmlReport != null;
     }
 
     private InputStream getXmlInputStream(File file) {

